python/cudf_polars/cudf_polars/experimental/groupby.py

```diff
@@ -0,0 +1,304 @@
+    ir: GroupBy, rec: LowerIRTransformer
+) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+    # Extract child partitioning
+    child, partition_info = rec(ir.children[0])
+
+    # Handle single-partition case
+    if partition_info[child].count == 1:
+        single_part_node = ir.reconstruct([child])
+        partition_info[single_part_node] = partition_info[child]
+        return single_part_node, partition_info
+
+    # Check group-by keys
+    if not all(expr.is_pointwise for expr in traversal([e.value for e in ir.keys])):
+        raise NotImplementedError(
+            f"GroupBy does not support multiple partitions for keys:\n{ir.keys}"
+        )  # pragma: no cover
+
+    # Check if we are dealing with any high-cardinality columns
+    post_aggregation_count = 1  # Default tree reduction
+    groupby_key_columns = [ne.name for ne in ir.keys]
+    cardinality_factor = {
+        c: min(f, 1.0)
+        for c, f in ir.config_options.get("executor_options", {})
+        .get("cardinality_factor", {})
+        .items()
+        if c in groupby_key_columns
+    }
+    if cardinality_factor:
+        # The `cardinality_factor` dictionary can be used
+        # to specify a mapping between column names and
+        # cardinality "factors". Each factor estimates the
+        # fractional number of unique values in the column.
+        # Each value should be in the range (0, 1].
+        child_count = partition_info[child].count
+        post_aggregation_count = max(
+            int(max(cardinality_factor.values()) * child_count),
+            1,
+        )
+
+    # Decompose the aggregation requests into three distinct phases
+    piecewise_exprs, reduction_exprs, selection_exprs = (
+        list(itertools.chain.from_iterable(x))
+        for x in zip(
+            *map(
+                partial(decompose_groupby_reduction, ir.schema),
+                ir.agg_requests,
+            ),
+            strict=False,
+        )
+    )
+
+    # Partition-wise groupby operation
+    gb_pwise = GroupBy(
+        ir.schema,
+        ir.keys,
+        piecewise_exprs,
+        ir.maintain_order,
+        ir.options,
+        ir.config_options,
+        child,
+    )
+    child_count = partition_info[child].count
+    partition_info[gb_pwise] = PartitionInfo(count=child_count)
+
+    # Add Shuffle node if necessary
+    gb_inter: GroupBy | Shuffle = gb_pwise
+    if post_aggregation_count > 1:
+        shuffle_options: dict[str, Any] = {}
+        gb_inter = Shuffle(
+            ir.schema,
+            ir.keys,
+            shuffle_options,
+            gb_pwise,
+        )
+        partition_info[gb_inter] = PartitionInfo(count=post_aggregation_count)
+
+    # Tree reduction if post_aggregation_count==1
+    # (Otherwise, this is another partition-wise op)
+    gb_reduce = GroupBy(
+        ir.schema,
+        ir.keys,
+        reduction_exprs,
+        ir.maintain_order,
+        ir.options,
+        ir.config_options,
+        gb_inter,
+    )
```

Comment:
*issue*: This reuse of the `ir.schema` for the partitionwise groupby and reduced groupby is incorrect I think. It should be advertising the schema of the intermediate piecewise exprs and reduction exprs.

So for the first one, I think the schema is `{k.name: k.dtype for k in ir.keys} | {k.name: k.dtype for k in piecewise_exprs}` and for the latter replace `piecewise_exprs` with `reduction_exprs`.

Reply:
Gah, that's true. I'll try to clean this up.